src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -60,128 +50,121 @@
         }
 
         { // read length encoding:
-            getOptimalIntegerEncoding(header, EncodingKey.RL_ReadLength, 0, records);
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createRANS(RANS.ORDER.ONE));
+            header.encodingMap.put(EncodingKey.RL_ReadLength, ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // alignment offset:
-            if (sorted) { // alignment offset:
+            if (sorted) {
                 header.APDelta = true;
-                getOptimalIntegerEncoding(header, EncodingKey.AP_AlignmentPositionOffset, 0, records);
+
+                final int aStartID = exCounter++;
+                header.encodingMap.put(EncodingKey.AP_AlignmentPositionOffset,
+                        ExternalIntegerEncoding.toParam(aStartID));
+                header.externalIds.add(aStartID);
+                header.externalCompressors.put(aStartID, ExternalCompressor.createRANS(RANS.ORDER.ZERO));
             } else {
                 final int aStartID = exCounter++;
                 header.APDelta = false;
                 header.encodingMap.put(EncodingKey.AP_AlignmentPositionOffset,
                         ExternalIntegerEncoding.toParam(aStartID));
                 header.externalIds.add(aStartID);
-                header.externalCompressors.put(aStartID,
-                        ExternalCompressor.createRANS(RANS.ORDER.ONE));
+                header.externalCompressors.put(aStartID, ExternalCompressor.createRANS(RANS.ORDER.ZERO));
                 log.debug("Assigned external id to alignment starts: " + aStartID);
             }
         }
 
         { // read group
-            getOptimalIntegerEncoding(header, EncodingKey.RG_ReadGroup, -1, records);
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createRANS(RANS.ORDER.ONE));
+            header.encodingMap.put(EncodingKey.RG_ReadGroup, ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // read name encoding:
-            final HuffmanParamsCalculator calculator = new HuffmanParamsCalculator();
-            for (final CramCompressionRecord record : records)
-                calculator.add(record.readName.length());
-            calculator.calculate();
-
-            header.encodingMap.put(EncodingKey.RN_ReadName, ByteArrayLenEncoding.toParam(
-                    HuffmanIntegerEncoding.toParam(calculator.values(),
-                            calculator.bitLens()), ExternalByteArrayEncoding
-                            .toParam(readNameID)));
+            header.externalIds.add(readNameID);
+            header.externalCompressors.put(readNameID, ExternalCompressor.createGZIP());
+            header.encodingMap.put(EncodingKey.RN_ReadName, ByteArrayStopEncoding.toParam((byte) '\t', readNameID));
```

Comment:
It looks `like CompressionHeader.addExternalCompressor` should maybe actually be 2 methods to account for the two different tag encodings.  What's the rational between choosing `ByteArrayStopEncoding` vs `ExternalIntegerEncoding`?